streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java

```diff
@@ -431,12 +407,97 @@ public void flush() {
             }
         };
         final StreamsMetrics streamsMetrics = new MockStreamsMetrics(new Metrics());
-        final StreamTask streamTask = new StreamTask(new TaskId(0, 0), "appId", partitions, topology, consumer,
-            restoreStateConsumer, createConfig(baseDir), streamsMetrics, stateDirectory, new ThreadCache("testCache", 0, streamsMetrics), new MockTime(), recordCollector);
+        final StreamTask streamTask = new StreamTask(taskId00, "appId", partitions, topology, consumer,
+                                                     restoreStateConsumer, createConfig(baseDir), streamsMetrics,
+                                                     stateDirectory, testCache, time, recordCollector);
         streamTask.flushState();
         assertTrue(flushed.get());
     }
 
+    @Test
+    public void shouldThrowIllegalStateExceptionIfCurrentNodeIsNotNullWhenPunctuateCalled() throws Exception {
+        ((ProcessorContextImpl) task.processorContext()).setCurrentNode(processor);
+        try {
+            task.punctuate(processor, 10);
+            fail("Should throw illegal state exception as current node is not null");
+        } catch (final IllegalStateException e) {
+            // pass
+        }
+    }
+
+    @Test
+    public void shouldCallPunctuateOnPassedInProcessorNode() throws Exception {
+        task.punctuate(processor, 5);
+        assertThat(processor.punctuatedAt, equalTo(5L));
+        task.punctuate(processor, 10);
+        assertThat(processor.punctuatedAt, equalTo(10L));
+    }
+
+    @Test
+    public void shouldSetProcessorNodeOnContextBackToNullAfterSuccesfullPunctuate() throws Exception {
+        task.punctuate(processor, 5);
+        assertThat(((ProcessorContextImpl) task.processorContext()).currentNode(), nullValue());
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void shouldThrowIllegalStateExceptionOnScheduleIfCurrentNodeIsNull() throws Exception {
+        task.schedule(1);
+    }
+
+    @Test
+    public void shouldNotThrowIExceptionOnScheduleIfCurrentNodeIsNotNull() throws Exception {
+        ((ProcessorContextImpl) task.processorContext()).setCurrentNode(processor);
+        task.schedule(1);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test
+    public void shouldThrowExceptionIfAnyExceptionsRaisedDuringCloseTopology() throws Exception {
+        task = createTaskThatThrowsExceptionOnClose();
```

Comment:
Should we close the task first before re-initialize it to another StreamTask? Ditto below.